src/main/java/org/truffleruby/language/supercall/LookupSuperMethodNode.java

```diff
@@ -82,4 +90,27 @@ protected int getCacheLimit() {
         return getContext().getOptions().METHOD_LOOKUP_CACHE;
     }
 
+    private DeclarationContext getDeclarationContext(VirtualFrame frame, InternalMethod currentMethod,
+            DynamicObject selfMetaClass) {
+        final DeclarationContext context = RubyArguments.getDeclarationContext(frame);
+
+        if (currentMethod.isRefined()) {
+            // super from the refined method has access to the parent's active refinements for the selfMetaClass
+            final DynamicObject[] classRefinements = currentMethod
+                    .getActiveRefinements()
+                    .getRefinementsFor(selfMetaClass);
+
+            if (classRefinements == null) {
+                return context;
+            } else {
+                // add to the context active refinements for the selfMetaClass
+                final Map<DynamicObject, DynamicObject[]> newRefinements = new HashMap<>(context.getRefinements());
+                newRefinements.put(selfMetaClass, classRefinements);
```

Comment:
That seems quite expensive. Do we need to look into the `RubyArguments.getDeclarationContext(frame)` or just `classRefinements` is enough to find the super method?

If we need it, I don't see much other way, so probably fine.
Would be worth checking if `context.getRefinements()` is empty then, in which case we don't need a new Map.

Reply:
> Do we need to look into the RubyArguments.getDeclarationContext(frame)

Yeah, we need it because it contains the method refinements:

```ruby
module A
  using SomeM
  
  def foo
    super + ??? # here we should see SomeM; aka RubyArguments.getDeclarationContext(frame)
  end
end

refine C
  include A # pass classRefinements
end
```

However, we know that `RubyArguments.getDeclarationContext(frame) ==  currentMethod.getDeclarationContext()` (it what RubyArguments.pack does) so we can replace one with another.

Additionally, it helps to remove `frame` arguments and return `@TruffleBoundary` 

See: https://github.com/oracle/truffleruby/pull/2039/commits/d1cdef8ff843f3e87d0a07449b7ac9a3d8e02561

Reply:
Do we have a spec that tests this, that `super` will also use refinements in its lexical scope (not just refinements when looking up the method which has `super` in it)?
Semantically it might make more sense to consider only one set of refinements for `super`, so from the first call, it's a well-defined way up the chain. But of course we should match MRI.

Reply:
Done: https://github.com/oracle/truffleruby/pull/2039/commits/05e27ecd9dda8891d184c553fdaea7bb03e5c61c